### ISSUE TITLE
SDXL 1.0 Inpainting - Lower result quality with certain masks

### ISSUE THREAD

**Issue by Madot10** [huggingface/diffusers#4392]
Hi there! I've been conducting tests using the sd-xl-1-0-base model with the pipeline `StableDiffusionXLInpaintPipeline`. 

However, I've noticed that the model generates results where the "white zone" doesn't seem to have any relation to the rest of the image. It appears as if another image has been pasted over the original.

![download](https://github.com/huggingface/diffusers/assets/30395090/4e4dcc4c-ff3e-4649-819b-6d5dce738497)

Notice how the line where the mask ends doesn't appear when using the SD 2 model

Is there a planned release for an inpainting model? Or do you recommend fine-tuning one myself?

**Comment by AmericanPresidentJimmyCarter** [comment]
> Is there a planned release for an inpainting model?

Dunno, the last one for SD 1/2 weights was from RunwayML and took tens of thousands of A100 hours.

> do you recommend fine-tuning one myself?

Not really unless you have tens of thousands of A100 hours at your disposal.

**Comment by sayakpaul** [comment]
Cc: @patrickvonplaten

**Comment by MoonRide303** [comment]
Same observation here - SDXL base model is not good enough for inpainting. I used sample images from [SDXL documentation](https://huggingface.co/docs/diffusers/api/pipelines/stable_diffusion/stable_diffusion_xl#inpainting), and "an empty bench" prompt. Results from sd-v1-5-inpainting model:
![image](https://github.com/huggingface/diffusers/assets/130458190/0d2cc119-234d-4657-a518-6d9c8dd6802a)

and output from sd_xl_base_1.0:
![image](https://github.com/huggingface/diffusers/assets/130458190/696053e2-3534-4954-a879-bb7510dfb473)

**Comment by patrickvonplaten** [comment]
Thanks for the interesting results here! Could you copy-paste some code snippets that you ran that gave these results? :-)

**Comment by bmc-synth** [comment]
Hi. I'm seeing similar issues, I haven't been able to get inpainting to give decent results, with or without refiner. For example:

```from diffusers import StableDiffusionXLInpaintPipeline
from diffusers.utils import load_image
import torch

pipe = StableDiffusionXLInpaintPipeline.from_pretrained(
    "stabilityai/stable-diffusion-xl-base-1.0", 
    torch_dtype=torch.float16, 
    variant="fp16", 
    use_safetensors=True,
)
pipe.to("cuda")

img_url = "https://raw.githubusercontent.com/CompVis/latent-diffusion/main/data/inpainting_examples/overture-creations-5sI6fQgYIuo.png"
mask_url = "https://raw.githubusercontent.com/CompVis/latent-diffusion/main/data/inpainting_examples/overture-creations-5sI6fQgYIuo_mask.png"

init_image = load_image(img_url)
mask_image = load_image(mask_url)

init_image.save("init_image.png")
mask_image.save("mask_image.png")

image = pipe(
    prompt="a dog",
    image=init_image,
    mask_image=mask_image,
    num_inference_steps=20,
).images
image[0].save("output.png")
```

![output](https://github.com/huggingface/diffusers/assets/99252446/e30420b0-fe1e-4b73-96f2-3776e77535c3)

**Comment by Madot10** [comment]
> Thanks for the interesting results here! Could you copy-paste some code snippets that you ran that gave these results? :-)

Hi!, sure, here is:
```
torch.backends.cuda.matmul.allow_tf32 = True
torch_device = "cuda" if torch.cuda.is_available() else "cpu"

pipe = StableDiffusionXLInpaintPipeline.from_pretrained("stabilityai/stable-diffusion-xl-base-1.0", torch_dtype=torch.float16, variant="fp16", use_safetensors=True) 

pipe = pipe.to(torch_device)
pipe.enable_xformers_memory_efficient_attention()


prompt = processed_data['prompt']
negative_prompt = processed_data['negative_prompt']

with torch.inference_mode():
        result = pipe(prompt=prompt, 
                     image=processed_data['init_image'],
                     mask_image = processed_data['mask_image'],
                     generator=generator, 
                     num_inference_steps=processed_data['steps'],
                     negative_prompt=negative_prompt,
                     height = processed_data['height'] if 'height' in processed_data else None,
                     width = processed_data['width'] if 'width' in processed_data else None,
                     guidance_scale=processed_data['guidance_scale'])
        
        image = result.images[0]
```

**Comment by bmc-synth** [comment]
I see. Could we somehow use one of the old inpaint models with the new refiner?

**Comment by MoonRide303** [comment]
@bmc-synth You can use base and/or refiner to further process any kind of image, if you go through img2img (out of latent space) and proper denoising control.

Sample workflow for ComfyUI below - picking up pixels from SD 1.5 inpainting model, and separately processing it (with different prompts) by both SDXL base and refiner models:
![image](https://github.com/huggingface/diffusers/assets/130458190/c823046a-8913-4823-9dee-794a4e567493)

**Comment by nasy** [comment]
I am experiencing the same issue. Does anyone know if Stability AI is planning on improving the inpainting for SDXL to fix this issues?

**Comment by patrickvonplaten** [comment]
Hmm, such an inpainting task with "chained" pipelines actually works really well with `diffusers`! I think the main problem here that we don't do a good job in our docs explaining how it should be done.

We really need to write a longer, good guide about all:
- text2img
- img2img
- inpainting

Opened an issue here: https://github.com/huggingface/diffusers/issues/4758

To give an example of how you could solve the above task in diffusers have a look here:

```py
from diffusers import AutoPipelineForInpainting, AutoPipelineForImage2Image
from diffusers.utils import load_image
import torch
import os

torch.backends.cuda.matmul.allow_tf32 = True
torch_device = "cuda" if torch.cuda.is_available() else "cpu"

pipe = AutoPipelineForInpainting.from_pretrained("runwayml/stable-diffusion-inpainting", torch_dtype=torch.float16, variant="fp16", use_safetensors=True) 

pipe = pipe.to(torch_device)
pipe.enable_xformers_memory_efficient_attention()

img_url = "https://raw.githubusercontent.com/CompVis/latent-diffusion/main/data/inpainting_examples/overture-creations-5sI6fQgYIuo.png"
mask_url = "https://raw.githubusercontent.com/CompVis/latent-diffusion/main/data/inpainting_examples/overture-creations-5sI6fQgYIuo_mask.png"

image = load_image(img_url)
mask_image = load_image(mask_url)

prompt = "dslr photography of an empty bench, high quality"
generator = torch.Generator(device="cuda").manual_seed(0)

image = pipe(
    prompt=prompt,
    image=image,
    mask_image=mask_image,
    guidance_scale=8.0,
    num_inference_steps=20,
    generator=generator,
).images[0]


image = image.resize((1024, 1024))

pipe = AutoPipelineForInpainting.from_pretrained("stabilityai/stable-diffusion-xl-refiner-1.0", torch_dtype=torch.float16, variant="fp16", use_safetensors=True) 
pipe.to("cuda")

pipe.enable_xformers_memory_efficient_attention()

image = pipe(
    prompt=prompt,
    image=image,
    mask_image=mask_image,
    guidance_scale=8.0,
    num_inference_steps=100,
    strength=0.2,
    generator=generator,
).images[0]

pipe = AutoPipelineForImage2Image.from_pipe(pipe)
pipe.enable_xformers_memory_efficient_attention()

image = pipe(
    prompt=prompt,
    image=image,
    guidance_scale=8.0,
    num_inference_steps=100,
    strength=0.2,
    generator=generator,
).images[0]
image
```

which gives:
![aaa (8)](https://github.com/huggingface/diffusers/assets/23423619/8bf4c274-10b7-4275-8d3a-61106f513275)

See google colab here: https://colab.research.google.com/drive/1IxYOiGHdiqBJfb-N7WXaSxmPkuNfdG1g?usp=sharing

**Comment by Madot10** [comment]
Hi! Thanks for the explanation! @patrickvonplaten 

It works 😄

I have some questions about what's happening behind the scenes. 
I'm assuming that the inpainting 1.5 pipeline processes the image first. Then, the output is passed to the inpainting XL pipeline which uses the refiner model to convert the image into a compatible latent format for the final pipeline (img2img XL).

In this scenario, is the Refiner model primarily used for making minor improvements to the image rather than a complete inpainting process in XL?

**Comment by andysingal** [comment]
> Hi. I'm seeing similar issues, I haven't been able to get inpainting to give decent results, with or without refiner. For example:
> 
> ```
> from diffusers.utils import load_image
> import torch
> 
> pipe = StableDiffusionXLInpaintPipeline.from_pretrained(
>     "stabilityai/stable-diffusion-xl-base-1.0", 
>     torch_dtype=torch.float16, 
>     variant="fp16", 
>     use_safetensors=True,
> )
> pipe.to("cuda")
> 
> img_url = "https://raw.githubusercontent.com/CompVis/latent-diffusion/main/data/inpainting_examples/overture-creations-5sI6fQgYIuo.png"
> mask_url = "https://raw.githubusercontent.com/CompVis/latent-diffusion/main/data/inpainting_examples/overture-creations-5sI6fQgYIuo_mask.png"
> 
> init_image = load_image(img_url)
> mask_image = load_image(mask_url)
> 
> init_image.save("init_image.png")
> mask_image.save("mask_image.png")
> 
> image = pipe(
>     prompt="a dog",
>     image=init_image,
>     mask_image=mask_image,
>     num_inference_steps=20,
> ).images
> image[0].save("output.png")
> ```
> 
> ![output](https://user-images.githubusercontent.com/99252446/258402351-e30420b0-fe1e-4b73-96f2-3776e77535c3.png)

I see the dog legs going in the bench… I guess it more has to do with the mask and image size … please use 1024x1024 for better results?

**Comment by JohanHuynh0130** [comment]
Hi everyone! Thank you for solving the issue! @patrickvonplaten I really appreciate it! I'm just wondering whether when you think SAI will release the SDXL inpainting model like what they did with 1.5?

Thanks!

**Comment by patrickvonplaten** [comment]
cc @williamberman

**Comment by ClementPinard** [comment]
Had a go with this finetuning : https://huggingface.co/diffusers/stable-diffusion-xl-1.0-inpainting-0.1

And it worked quite well.

As stated in the "limitations" tab, it's not perfect when strength is set to 1, and a strength of e.g. 0.99 can let the removed bit alter the result.

For example, I wanted to replace a blue car with a red truck, I had a truck, but it stayed blue.

What I ended up doing is similar to solutions presented above, which is doing the inpainting with SD2, and then again with SDXL, with a strength of 0.99
